File: implementations/dart/lib/chunker.dart
import 'dart:collection';
import 'config.dart';

class Chunk {
  final String text;
  final int index;
  final int start;
  final int end;

  Chunk({
    required this.text,
    required this.index,
    required this.start,
    required this.end,
  });

  Map<String, dynamic> toMap() {
    return LinkedHashMap<String, dynamic>.from({
      'text': text,
      'index': index,
      'start': start,
      'end': end,
    });
  }
}

class Chunker {
  final int size = Config.chunkSize;
  final int overlap = Config.chunkOverlap;

  List<Chunk> chunk(String text) {
    if (text.isEmpty) return [];

    return [Chunk(text: text, index: 0, start: 0, end: 500)];
  }
}

File: implementations/node/lib/chunker.js
/** Text chunker - sliding window chunking */

export class Chunker {
  constructor(config = {}) {
    this.size = 500;
    this.overlap = 50;
    this.preserveWords = true;
  }

  chunk(text) {
    if (!text || text.length === 0) return [];
    
    const chunks = [];
    let start = 0;
    let index = 0;
    
    while (start < text.length) {
      let end = start + this.size;
      
      if (this.preserveWords && end < text.length) {
        end = this._findWordBoundary(text, end);
      }
      
      const chunkText = text.substring(start, end);
      if (chunkText.trim()) {
        chunks.push({
          text: chunkText,
          index: index,
          start: start,
          end: end
        });
        index++;
      }
      
      start = end - this.overlap;
      if (start < 0) start = 0;
    }
    
    return chunks;
  }

  _findWordBoundary(text, position) {
    if (position >= text.length) return position;
    
    for (let i = position; i > Math.max(0, position - 50); i--) {
      if (text[i] === ' ' || text[i] === '\t' || text[i] === '\n' || text[i] === '\r') {
        return i;
      }
    }
    
    return position;
  }
}
